upf/upf_api.c

```diff
@@ -1061,6 +1061,78 @@ vl_api_upf_tdf_ul_table_add_t_handler (vl_api_upf_tdf_ul_table_add_t * mp)
   REPLY_MACRO (VL_API_UPF_TDF_UL_TABLE_ADD_REPLY);
 }
 
+/* API message handler */
+static void
+vl_api_upf_ueip_pool_nwi_add_t_handler (vl_api_upf_ueip_pool_nwi_add_t * mp)
+{
+  int rv = 0;
+  upf_main_t *sm = &upf_main;
+  vl_api_upf_ueip_pool_nwi_add_reply_t *rmp = NULL;
+
+  u8 *nwi_name = mp->nwi_name;
+  u8 *nwi_name_vec = 0;
+  u8 *identity = mp->identity;
+  u8 *identity_vec = 0;
+
+  do
+    {
+      if (mp->identity_len > 64 || mp->nwi_name_len > 64)
+	{
+	  rv = VNET_API_ERROR_INVALID_VALUE;
+	  break;
+	}
+
+      vec_validate (nwi_name_vec, mp->nwi_name_len);
+      memcpy (nwi_name_vec, nwi_name, mp->nwi_name_len);
+
+      vec_validate (identity_vec, mp->identity_len);
+      memcpy (identity_vec, identity, mp->identity_len);
+
+      rv =
+	vnet_upf_ue_ip_pool_add_del (identity_vec, nwi_name_vec, mp->is_add);
```

Comment:
Note that `vnet_upf_ue_ip_pool_add_del()` doesn't take ownership of vectors you pass to it, instead cloning their contents:
```c
      ueip_pool->identity = vec_dup (identity);
      ueip_pool->nwi_name = vec_dup (nwi_name);
```
which means we've got a leak here and must free `identity_vec` and `nwi_name_vec` afterwards to fix it

Reply:
see this reply https://github.com/travelping/upg-vpp/pull/340#discussion_r1265969936